### PR TITLE
Solve: Inconsistent decoding with additional special tokens between slow and fast tokenizers.

### LINKED ISSUE #28287: [MBart50] Inconsistent decoding with additional special tokens between slow and fast tokenizers
### System Info

- `transformers` version: 4.36.2
- Platform: Linux-6.2.0-25-generic-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.1
- Safetensors version: 0.4.1
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): not installed (NA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker @younesbelkada 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Load a non-fast Tokenizer for mBART
2. Add an additional special token to it
3. Encode and then decode input containing previously added special token

```python3
from transformers import MBart50Tokenizer

tk = MBart50Tokenizer.from_pretrained('facebook/mbart-large-50')
tk.add_tokens('<token>', True)
print(tk.decode(tk("This is my example sentence with a special <token> token")["input_ids"]))
>>> 'en_XXThis is my example sentence with a special <token> token</s>'
```
This differs from the fast tokenizers' decoding scheme, as it will correctly decode the input with a space after `en_XX`. I believe this is due to the implementation for `legacy_added_tokens` in https://github.com/huggingface/transformers/blob/3cefac1d974db5e2825a0cb2b842883a628be7a0/src/transformers/tokenization_utils.py#L1002-L1022
and more specifically the second part of the set definition for `legacy_added_tokens` that accounts for special tokens that have been added manually after loading (?)

When disabling the special handling for `legacy_added_tokens`, the tokenization output would be correct, so I was primarily wondering for what reason this was added and whether removing this would potentially break other tokenizers.

### Expected behavior

```python3
fast_tk = MBart50TokenizerFast.from_pretrained('facebook/mbart-large-50')
fast_tk.add_tokens('<token>', True)
print(fast_tk.decode(fast_tk("This is my example sentence with a special <token> token")["input_ids"])))
>>> 'en_XX This is my example sentence with a special <token> token</s>'
```
The decoding should match the fast tokenizers' output (?), at least I would assume so.

### PR DESCRIPTION
**Pull request by hisushanta**
With this pull request, I have endeavored to remedy a minor decoding issue. If any problems remain with my proposed solution, I welcome your thoughtful feedback and suggestions for improvement.

Fixes #28287 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@ArthurZucker   @amyeroberts

### CODE REVIEW

File: src/transformers/models/mbart50/tokenization_mbart50.py [hunk 1]
```diff
@@ -246,7 +246,7 @@ def convert_tokens_to_string(self, tokens):
             else:
                 current_sub_tokens.append(token)
                 prev_is_special = False
-        out_string += self.sp_model.decode(current_sub_tokens)
+        out_string += " " + self.sp_model.decode(current_sub_tokens)
```
Comment (ArthurZucker) [inline src/transformers/models/mbart50/tokenization_mbart50.py:249]: I am not 100% convinced that this is the fix! 
The reason is that if you always add this, then normal decoding will always add this, which is not expected. 
This should only be added if the previous token is special in the decoding functions. Make sure to add a test based on the issue 🤗

Comment (hi-sushanta) [reply src/transformers/models/mbart50/tokenization_mbart50.py]: Please now check if it's ok or not.

**Comment by hi-sushanta** [comment]
Would you be able to check my solution?

File: .circleci/create_circleci_config.py [hunk 2]
```diff
@@ -517,7 +517,7 @@ def job_name(self):
         "pip install -U --upgrade-strategy eager 'natten<0.15.0'",
         "pip install -U --upgrade-strategy eager g2p-en",
         "find -name __pycache__ -delete",
-        "find . -name \*.pyc -delete",
+        r"find . -name \*.pyc -delete",
```
Comment (amyeroberts) [inline .circleci/create_circleci_config.py:520]: Why this change? The circle CI configs shouldn't be touched

File: scripts/benchmark/trainer-benchmark.py [hunk 3]
```diff
@@ -186,11 +186,11 @@ def get_base_command(args, output_dir):
     args.base_cmd = re.sub(r"[\\\n]+", " ", args.base_cmd)
 
     # remove --output_dir if any and set our own
-    args.base_cmd = re.sub("--output_dir\s+[^\s]+", "", args.base_cmd)
+    args.base_cmd = re.sub(r"--output_dir\s+[^\s]+", "", args.base_cmd)
```
Comment (amyeroberts) [inline scripts/benchmark/trainer-benchmark.py:189]: Same here - why these changes in this file?

File: tests/models/mbart50/test_tokenization_mbart50.py [hunk 4]
```diff
@@ -304,3 +304,11 @@ def test_tokenizer_translation(self):
                 "forced_bos_token_id": 250001,
             },
         )
+
+    def test_inconsistent_decoding(self):
+        text = "<s>[special_token] normal text"
```
Comment (amyeroberts) [inline tests/models/mbart50/test_tokenization_mbart50.py:309]: I don't think this tests special tokens at all - writing `[special_token]` doesn't make it a special token.

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on a fix! 

There's several unrelated changes in this PR which will need to be removed before merging. Also, the test doesn't test tokenization on special tokens at all.

File: scripts/benchmark/trainer-benchmark.py (hunk 3 above)
Comment (hi-sushanta) [reply scripts/benchmark/trainer-benchmark.py]: I don't know why these codes are changed. but when I run the ruff command that time it's automatically added.

Comment (hi-sushanta) [reply scripts/benchmark/trainer-benchmark.py]: Now I solved it.

File: tests/models/mbart50/test_tokenization_mbart50.py (hunk 4 above)
Comment (hi-sushanta) [reply tests/models/mbart50/test_tokenization_mbart50.py]: Please now check if it's ok.

Comment (amyeroberts) [reply tests/models/mbart50/test_tokenization_mbart50.py]: The space is eaten between `special` and `<unk>` which I don't believe is intended. You also are not testing for the two cases as requested by @ArthurZucker in [his comment](https://github.com/huggingface/transformers/pull/28424#discussion_r1448968133).

**Comment by amyeroberts** [comment]
@hi-sushanta To figure out the errors, you'll need to look at the CI runs and debug locally so that the currently failing tests pass. For the quality checks, you'll need to fun `make fix-copies` and `make fixup`

**Comment by hi-sushanta** [comment]
But this command returns to the original codebase.
